internal/app/machined/pkg/controllers/network/device_config.go

```diff
@@ -156,13 +169,48 @@ func (ctrl *DeviceConfigController) Run(ctx context.Context, r controller.Runtim
 	}
 }
 
-func (ctrl *DeviceConfigController) getDeviceBySelector(device talosconfig.Device, links []resource.Resource) error {
+func (ctrl *DeviceConfigController) getDeviceBySelector(device talosconfig.Device, links safe.List[*network.LinkStatus]) error {
 	selector := device.Selector()
 
-	for _, link := range links {
-		linkStatus := link.(*network.LinkStatus).TypedSpec() //nolint:forcetypeassert,errcheck
+	matches := ctrl.selectDevices(selector, links)
+	if len(matches) == 0 {
+		return fmt.Errorf("no matching network device for defined selector: %v", selector)
+	}
+
+	link := matches[0]
+
+	dev := device.(*v1alpha1.Device) //nolint:errcheck,forcetypeassert
+	dev.DeviceInterface = link.Metadata().ID()
+
+	return nil
+}
+
+func (ctrl *DeviceConfigController) expandBondSelector(device talosconfig.Device, links safe.List[*network.LinkStatus]) error {
```

Comment:
Since are type asserting to `*v1alpha1.Device` anyway in `Run` maybe it makes sense to pass the concrete type here?